sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -767,29 +745,201 @@ private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath)
         }
     }
 
-    private void downloadToFileCleanup(AsynchronousFileChannel channel) {
-        try {
-            channel.close();
-        } catch (IOException e) {
-            throw logger.logExceptionAsError(new UncheckedIOException(e));
-        }
+    private Mono<Response<BlobProperties>> downloadToFileImpl(AsynchronousFileChannel file, BlobRange rangeReal,
+        ParallelTransferOptions finalParallelTransferOptions, DownloadRetryOptions downloadRetryOptions,
+        BlobRequestConditions requestConditions, boolean rangeGetContentMd5, Context context) {
+        // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
+        Lock progressLock = new ReentrantLock();
+        AtomicLong totalProgress = new AtomicLong(0);
+
+        /*
+         * Downloads the first chunk and gets the size of the data and etag if not specified by the user.
+         */
+       return getSetupMono(rangeReal, finalParallelTransferOptions, downloadRetryOptions, requestConditions,
+           rangeGetContentMd5, context)
+        .flatMap(setupTuple3 -> {
+            long newCount = setupTuple3.getT1();
+            BlobRequestConditions realConditions = setupTuple3.getT2();
+
+            int numChunks = calculateNumBlocks(newCount, finalParallelTransferOptions.getBlockSize());
+
+            // In case it is an empty blob, this ensures we still actually perform a download operation.
+            numChunks = numChunks == 0 ? 1 : numChunks;
```

Comment:
For the case when the blob is empty could we return `Mono.just(writeBodyToFile(initialResponse, file, 0, options, lock, progess)`?

Reply:
I think that should be functionally equivalent. Is this a code style suggestion you think will make the logic clearer?

Reply:
But I don't actually think this will clean anything up. I still need the special case for the first element to write the body to the file without calling download. I think this is the cleanest way to get the empty file case into the rest of the workflow.